sw/host/opentitanlib/src/test_utils/extclk.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Utilities for enabling or disabling external clock injection through JTAG to the clock manager.
+
+use std::time::Duration;
+
+use thiserror::Error;
+
+use top_earlgrey::top_earlgrey_memory;
+
+use crate::chip::boolean::MultiBitBool4;
+use crate::dif::clkmgr::{ClkmgrExtclkCtrl, ClkmgrReg};
+use crate::io::jtag::Jtag;
+use crate::test_utils::poll;
+
+/// Available speeds for the external clock.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ClockSpeed {
+    /// Low speed clock.
+    Low,
+    /// Nominal high speed clock.
+    High,
+}
+
+/// Controls for enabling or disabling external clock injection.
+pub struct ExternalClock;
+
+impl ExternalClock {
+    /// Base address of the clock manager.
+    const CLKMGR_BASE_ADDR: u32 = top_earlgrey_memory::TOP_EARLGREY_CLKMGR_AON_BASE_ADDR as u32;
+
+    /// Addresses of clock manager registers in memory.
+    const EXTCLK_CTRL_REGWEN_ADDR: u32 =
+        Self::CLKMGR_BASE_ADDR + ClkmgrReg::ExtclkCtrlRegwen as u32;
+    const EXTCLK_CTRL_ADDR: u32 = Self::CLKMGR_BASE_ADDR + ClkmgrReg::ExtclkCtrl as u32;
+    const EXTCLK_STATUS_ADDR: u32 = Self::CLKMGR_BASE_ADDR + ClkmgrReg::ExtclkStatus as u32;
+
+    /// Durations for polling of the STATUS register.
+    const POLL_TIMEOUT: Duration = Duration::from_millis(500);
+    const POLL_DELAY: Duration = Duration::from_millis(5);
+
+    /// Enable the external clock at the given speed.
+    pub fn enable(jtag: &dyn Jtag, speed: ClockSpeed) -> Result<(), ExternalClockError> {
+        Self::set_regwen(jtag, true)?;
+
+        // Fields and their values for the EXTCLK_CTRL register.
+        let hi_speed = match speed {
+            ClockSpeed::Low => MultiBitBool4::False,
+            ClockSpeed::High => MultiBitBool4::True,
+        };
+        let fields = [
+            (ClkmgrExtclkCtrl::SEL, MultiBitBool4::True),
+            (ClkmgrExtclkCtrl::HI_SPEED_SEL, hi_speed),
+        ];
+
+        // OR the fields together to get the register value.
+        let extclk_ctrl = fields.into_iter().fold(0, |acc, (field, value)| {
+            acc | field.emplace(u8::from(value) as u32)
+        });
+
+        // Set the EXTCLK_CTRL register.
+        jtag.write_memory32(Self::EXTCLK_CTRL_ADDR, &[extclk_ctrl])
+            .map_err(|source| ExternalClockError::Jtag { source })?;
+
+        Self::wait_for_ack(jtag)?;
+
+        // No longer need write access.
+        Self::set_regwen(jtag, false)?;
+
+        Ok(())
+    }
+
+    /// Disable the external clock.
+    pub fn disable(jtag: &dyn Jtag) -> Result<(), ExternalClockError> {
+        Self::set_regwen(jtag, true)?;
```

Comment:
The EXTCLK_CTRL_REGWEN is a rw0c (read / write-0-to-clear) CSR that resets to 1. Writing 1 to this register will do nothing. In other words, after power on reset, the register write enable is set to 1 by default, enabling writes to the EXTCLK_CTRL register. It is only if you clear the regwen (by writing 0 to it), will you not be able to write to the EXTCLK_CTRL CSR. You could instead read the REGWEN to check it is 1, and error out otherwise.

Reply:
Awesome, thanks
I'll swap this out for a `REGWEN` read